Applications/Piano/AudioEngine.h

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2019-2020, William McPherson <willmcpherson2@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "Music.h"
+#include <AK/FixedArray.h>
+#include <AK/Queue.h>
+
+class AudioEngine {
+public:
+    explicit AudioEngine();
+    ~AudioEngine();
+
+    Sample* buffer() const { return m_front_buffer_ptr; }
+    int octave() const { return m_octave; }
+    int octave_base() const { return (m_octave - octave_min) * 12; }
+    int wave() const { return m_wave; }
+    int decay() const { return m_decay; }
+    int delay() const { return m_delay; }
+    int time() const { return m_time; }
+    int tick() const { return m_tick; }
+
+    void fill_buffer(Sample* buffer);
+    void set_note(int note, Switch);
+    void set_note_current_octave(int note, Switch);
+    void set_octave(Direction);
+    void set_wave(int wave);
+    void set_wave(Direction);
+    void set_decay(int decay);
+    void set_delay(int delay);
+
+private:
+    double sine(size_t note);
+    double saw(size_t note);
+    double square(size_t note);
+    double triangle(size_t note);
+    double noise() const;
+
+    Sample m_front_buffer[sample_count] { 0, 0 };
+    Sample m_back_buffer[sample_count] { 0, 0 };
+    Sample* m_front_buffer_ptr { m_front_buffer };
+    Sample* m_back_buffer_ptr { m_back_buffer };
```

Comment:
Is this class uncopyable and unmovable? Better make it such!

Reply:
Like this?
```
AudioEngine& operator=(const AudioEngine&) = delete;
AudioEngine(AudioEngine&&) = delete;
AudioEngine& operator=(AudioEngine&&) = delete;
```

Reply:
Like `AK_MAKE_NONCOPYABLE` and `AK_MAKE_NONMOVABLE` :smile: